File: aep-lib-go/pkg/openapi/openapi.go
package openapi

import (
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"os"
	"strings"
)

const (
	OAS2        = "2.0"
	OAS3        = "3.0"
	ContentType = "application/json"
)

type OpenAPI struct {
	// oas 2.0 has swagger in the root.
	Swagger    string               `json:"swagger,omitempty"`
	Info       Info                 `json:"info"`
	OpenAPI    string               `json:"openapi,omitempty"`
	Servers    []Server             `json:"servers,omitempty"`
	Paths      map[string]*PathItem `json:"paths"`
	Components Components           `json:"components,omitempty"`
	// oas 2.0 has definitions in the root.
	Definitions map[string]Schema `json:"definitions,omitempty"`
}

func (o *OpenAPI) OASVersion() string {
	if o.Swagger == "2.0" {
		return OAS2
	} else if o.OpenAPI != "" {
		return OAS3
	}
	return ""
}

func (o *OpenAPI) DereferenceSchema(schema Schema) (*Schema, error) {
	// we dereference only local schemas for now.
	// TODO: add support for external schemas (prefixed by //)
	if schema.Ref != "" && strings.HasPrefix(schema.Ref, "#components/schema") {
		parts := strings.Split(schema.Ref, "/")
		key := parts[len(parts)-1]
		var childSchema Schema
		var ok bool
		switch o.OASVersion() {
		case OAS2:
			childSchema, ok = o.Definitions[key]
			slog.Debug("oasv2.0", "key", key)
			if !ok {
				return nil, fmt.Errorf("schema %q not found", schema.Ref)
			}
		default:
			childSchema, ok = o.Components.Schemas[key]
			if !ok {
				return nil, fmt.Errorf("schema %q not found", schema.Ref)
			}
		}
		return o.DereferenceSchema(childSchema)
	}
	return &schema, nil
}

func (o *OpenAPI) GetSchemaFromResponse(r Response) *Schema {
	switch o.OASVersion() {
	case OAS2:
		return r.Schema
	default:
		ct := r.Content[ContentType]
		return ct.Schema
	}
}

func (o *OpenAPI) GetSchemaFromRequestBody(r RequestBody) *Schema {
	switch o.OASVersion() {
	case OAS2:
		return r.Schema
	default:
		ct := r.Content[ContentType]
		return ct.Schema
	}
}

type Contact struct {
	Name  string `json:"name,omitempty"`
	Email string `json:"email,omitempty"`
	URL   string `json:"url,omitempty"`
}

type Server struct {
	URL         string                    `json:"url"`
	Description string                    `json:"description,omitempty"`
	Variables   map[string]ServerVariable `json:"variables,omitempty"`
}

type ServerVariable struct {
	Enum        []string `json:"enum,omitempty"`
	Default     string   `json:"default"`
	Description string   `json:"description,omitempty"`
}

type Info struct {
	Title       string  `json:"title"`
	Description string  `json:"description"`
	Version     string  `json:"version"`
	Contact     Contact `json:"contact,omitempty"`
}

type PathItem struct {
	Get    *Operation `json:"get,omitempty"`
	Patch  *Operation `json:"patch,omitempty"`
	Post   *Operation `json:"post,omitempty"`
	Put    *Operation `json:"put,omitempty"`
	Delete *Operation `json:"delete,omitempty"`
}

type Operation struct {
	Summary                  string                    `json:"summary,omitempty"`
	Description              string                    `json:"description,omitempty"`
	OperationID              string                    `json:"operationId,omitempty"`
	Parameters               []Parameter               `json:"parameters,omitempty"`
	Responses                map[string]Response       `json:"responses,omitempty"`
	RequestBody              *RequestBody              `json:"requestBody,omitempty"`
	XAEPLongRunningOperation *XAEPLongRunningOperation `json:"x-aep-long-running-operation,omitempty"`
}

type Parameter struct {
	Name            string           `json:"name,omitempty"`
	In              string           `json:"in,omitempty"`
	Description     string           `json:"description,omitempty"`
	Required        bool             `json:"required,omitempty"`
	Schema          *Schema          `json:"schema,omitempty"`
	Type            string           `json:"type,omitempty"`
	XAEPResourceRef *XAEPResourceRef `json:"x-aep-resource-reference,omitempty"`
}

type Response struct {
	Description string               `json:"description,omitempty"`
	Content     map[string]MediaType `json:"content,omitempty"`
	// oas 2.0 has the schema in the response.
	Schema *Schema `json:"schema,omitempty"`
}

type RequestBody struct {
	Description string               `json:"description,omitempty"`
	Content     map[string]MediaType `json:"content"`
	Required    bool                 `json:"required"`
	// oas 2.0 has the schema in the request body.
	Schema *Schema `json:"schema,omitempty"`
}

type MediaType struct {
	Schema *Schema `json:"schema,omitempty"`
}

type Schema struct {
	Type            string        `json:"type,omitempty"`
	Format          string        `json:"format,omitempty"`
	Items           *Schema       `json:"items,omitempty"`
	Properties      Properties    `json:"properties,omitempty"`
	Ref             string        `json:"$ref,omitempty"`
	XAEPResource    *XAEPResource `json:"x-aep-resource,omitempty"`
	XAEPFieldNumber int           `json:"x-aep-field-number,omitempty"`
	// Documents the name of the proto message to use for generation.
	// If unset, proto generation will create a proto message for this schema.
	XAEPProtoMessageName string   `json:"x-aep-proto-message-name,omitempty"`
	ReadOnly             bool     `json:"readOnly,omitempty"`
	Required             []string `json:"required,omitempty"`
	Description          string   `json:"description,omitempty"`
	AdditionalProperties bool     `json:"additionalProperties,omitempty"`
}

type Properties map[string]Schema

type Components struct {
	Schemas map[string]Schema `json:"schemas"`
}

type XAEPResource struct {
	Singular string   `json:"singular,omitempty"`
	Plural   string   `json:"plural,omitempty"`
	Patterns []string `json:"patterns,omitempty"`
	Parents  []string `json:"parents,omitempty"`
}

type XAEPResourceRef struct {
	Resource string `json:"resource,omitempty"`
}

type XAEPLongRunningOperation struct {
	Response XAEPLongRunningOperationResponse `json:"response,omitempty"`
}

type XAEPLongRunningOperationResponse struct {
	Schema *Schema `json:"schema,omitempty"`
}

func FetchOpenAPI(pathOrURL string) (*OpenAPI, error) {
	body, err := readFileOrURL(pathOrURL)
	if err != nil {
		return nil, fmt.Errorf("unable to read file or URL: %w", err)
	}

	var api OpenAPI
	if err := json.Unmarshal(body, &api); err != nil {
		return nil, err
	}

	return &api, nil
}

func readFileOrURL(pathOrURL string) ([]byte, error) {
	if isURL(pathOrURL) {
		resp, err := http.Get(pathOrURL)
		if err != nil {
			return nil, err
		}
		defer resp.Body.Close()
		return io.ReadAll(resp.Body)
	}

	return os.ReadFile(pathOrURL)
}

func isURL(str string) bool {
	u, err := url.Parse(str)
	return err == nil && (u.Scheme == "http" || u.Scheme == "https")
}

File: aep-lib-ts/src/openapi/openapi.ts
import {
  OpenAPI,
  OAS2,
  OAS3,
  ContentType,
  Schema,
  Response,
  RequestBody,
} from "./types.js";

import yaml from "js-yaml";

export class OpenAPIImpl {
  private api: OpenAPI;

  constructor(api: OpenAPI) {
    this.api = api;
  }

  oasVersion(): string {
    if (this.api.swagger === "2.0") {
      return OAS2;
    } else if (this.api.openapi) {
      return OAS3;
    }
    return "";
  }

  async dereferenceSchema(schema: Schema): Promise<Schema> {
    if (schema.$ref) {
      const parts = schema.$ref.split("/");
      const key = parts[parts.length - 1];
      let childSchema: Schema | undefined;

      switch (this.oasVersion()) {
        case OAS2:
          childSchema = this.api.definitions?.[key];
          if (!childSchema) {
            throw new Error(`schema ${schema.$ref} not found`);
          }
          break;
        default:
          childSchema = this.api.components?.schemas[key];
          if (!childSchema) {
            throw new Error(`schema ${schema.$ref} not found`);
          }
      }
      return this.dereferenceSchema(childSchema);
    }
    return schema;
  }

  getSchemaFromResponse(response: Response): Schema | undefined {
    switch (this.oasVersion()) {
      case OAS2:
        return response.schema;
      default:
        return response.content?.[ContentType]?.schema;
    }
  }

  getSchemaFromRequestBody(requestBody: RequestBody): Schema | undefined {
    switch (this.oasVersion()) {
      case OAS2:
        return requestBody.schema;
      default:
        return requestBody.content[ContentType]?.schema;
    }
  }
}

export async function fetchOpenAPI(pathOrURL: string): Promise<OpenAPI> {
  try {
    const body = await readFileOrURL(pathOrURL);
    return parseYAML(body);
  } catch (error) {
    throw new Error(
      `Failed to fetch or parse OpenAPI schema from ${pathOrURL}: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
}

async function readFileOrURL(pathOrURL: string): Promise<string> {
  if (isURL(pathOrURL)) {
    const response = await fetch(pathOrURL);
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    return response.text();
  }

  // In Node.js environment
  if (typeof process !== "undefined" && process.versions?.node) {
    const fs = await import("fs/promises");
    return fs.readFile(pathOrURL, "utf-8");
  }

  throw new Error("File system access not available in this environment");
}

function isURL(str: string): boolean {
  try {
    const url = new URL(str);
    return url.protocol === "http:" || url.protocol === "https:";
  } catch {
    return false;
  }
}

async function parseYAML(yamlString: string): Promise<OpenAPI> {
  return yaml.load(yamlString) as OpenAPI;
}
